File: ios/Classes/controller/RtpSenderController.swift
import Flutter

/// Controller for the `RtpSender`.
class RtpSenderController {
  /// Flutter messenger for creating channels.
  private var messenger: FlutterBinaryMessenger

  /// Instance of `RtpSender` proxy.
  private var rtpSender: RtpSenderProxy

  /// ID of channels created for this controller.
  private var channelId: Int = ChannelNameGenerator.nextId()

  /// Method channel for communicating with Flutter side.
  private var channel: FlutterMethodChannel

  /// Creates a new `RtpSenderController` for the provided `RtpSenderProxy`.
  init(messenger: FlutterBinaryMessenger, rtpSender: RtpSenderProxy) {
    let channelName = ChannelNameGenerator.name(name: "RtpSender", id: channelId)
    self.messenger = messenger
    self.rtpSender = rtpSender
    self.channel = FlutterMethodChannel(name: channelName, binaryMessenger: messenger)
    self.channel.setMethodCallHandler({ (call, result) in
      self.onMethodCall(call: call, result: result)
    })
  }

  /// Handles all supported Flutter method calls for the `RtpSenderProxy`.
  func onMethodCall(call: FlutterMethodCall, result: FlutterResult) {
    let argsMap = call.arguments as? [String: Any]
    switch call.method {
    case "replaceTrack":
      let trackId = argsMap!["trackId"] as? String
      var track: MediaStreamTrackProxy?
      if trackId != nil {
        track = MediaStreamTrackStore.tracks[trackId!]
      } else {
        track = nil
      }

      self.rtpSender.replaceTrack(t: track)
      result(nil)
    case "dispose":
      self.channel.setMethodCallHandler(nil)
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  /// Converts this controller to the Flutter method call response.
  func asFlutterResult() -> [String: Any] {
    [
      "channelId": channelId
    ]
  }
}

File: android/src/main/kotlin/com/instrumentisto/medea_flutter_webrtc/controller/RtpSenderController.kt
package com.instrumentisto.medea_flutter_webrtc.controller

import com.instrumentisto.medea_flutter_webrtc.TrackRepository
import com.instrumentisto.medea_flutter_webrtc.proxy.RtpSenderProxy
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Controller of [RtpSenderProxy] functional.
 *
 * @param messenger Messenger used for creating new [MethodChannel]s.
 * @property sender Underlying [RtpSenderProxy] to perform [MethodCall]s on.
 */
class RtpSenderController(messenger: BinaryMessenger, private val sender: RtpSenderProxy) :
    MethodChannel.MethodCallHandler, IdentifiableController {
  /** Unique ID of the [MethodChannel] of this controller. */
  private val channelId = nextChannelId()

  /** Channel listened for the [MethodCall]s. */
  private val chan = MethodChannel(messenger, ChannelNameGenerator.name("RtpSender", channelId))

  init {
    chan.setMethodCallHandler(this)
  }

  override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
    when (call.method) {
      "replaceTrack" -> {
        val trackId: String? = call.argument("trackId")
        val track =
            if (trackId != null) {
              TrackRepository.getTrack(trackId)!!
            } else {
              null
            }
        sender.replaceTrack(track)
        result.success(null)
      }
      "encodings" -> {
        val res =
            sender.getParameters().encodings.mapIndexed { index, enc ->
              mapOf(
                  "index" to index,
                  "rid" to enc.rid,
                  "active" to enc.active,
                  "maxBitrate" to enc.maxBitrateBps,
                  "maxFramerate" to enc.maxFramerate,
                  "scaleResolutionDownBy" to enc.scaleResolutionDownBy,
              )
            }

        result.success(res)
      }
      "setParameters" -> {
        val params = sender.getParameters()
        val encodings: List<Map<String, Any>>? = call.argument("encodings")

        encodings?.forEach {
          val enc = params.encodings[it["index"] as Int]
          enc.active = it["active"] as Boolean
          enc.maxBitrateBps = it["maxBitrate"] as Int?
          enc.maxFramerate = it["maxFramerate"] as Int?
          enc.scaleResolutionDownBy = it["scaleResolutionDownBy"] as Double?
        }

        if (encodings != null && !sender.setParameters(params)) {
          result.error("SenderException", "Could not set parameters", null)
        }

        result.success(null)
      }
      "dispose" -> {
        chan.setMethodCallHandler(null)
        result.success(null)
      }
    }
  }

  /**
   * Converts this [RtpSenderController] to the Flutter's method call result.
   *
   * @return [Map] generated from this controller which can be returned to the Flutter side.
   */
  fun asFlutterResult(): Map<String, Any> = mapOf("channelId" to channelId)
}
